moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp

```diff
@@ -0,0 +1,93 @@
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of KU Leuven nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jeroen De Maeyer */
+
+#include <moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space.h>
+
+#include <ompl/base/spaces/constraint/ConstrainedStateSpace.h>
+
+const std::string ompl_interface::ConstrainedPlanningStateSpace::PARAMETERIZATION_TYPE =
+    "ConstrainedPlanningJointModel";
+
+ompl_interface::ConstrainedPlanningStateSpace::ConstrainedPlanningStateSpace(
+    const ModelBasedStateSpaceSpecification& spec)
+  : ModelBasedStateSpace(spec)
+{
+  setName(getName() + "_" + PARAMETERIZATION_TYPE);
+}
+
+double* ompl_interface::ConstrainedPlanningStateSpace::getValueAddressAtIndex(ompl::base::State* ompl_state,
+                                                                              const unsigned int index) const
+{
+  if (index >= variable_count_)
+    return nullptr;
+
+  // Developer tip: replace this with a dynamic_cast for debugging
+  return ompl_state->as<ompl_interface::ConstrainedPlanningStateSpace::StateType>()->values + index;
+}
+
+void ompl_interface::ConstrainedPlanningStateSpace::copyToRobotState(moveit::core::RobotState& robot_state,
+                                                                     const ompl::base::State* ompl_state) const
+{
+  robot_state.setJointGroupPositions(
+      spec_.joint_model_group_,
+      ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState()->as<StateType>()->values);
+  robot_state.update();
+}
```

Comment:
## Assert statement
I like the idea of adding `assert(ptr != nullptr);`. There are many other places in the existing OMPL interface where this could be valuable, but I feel like that does not belong in this PR. I'll just add it to the new code.

## `state->as<StateType>()` pattern
This pattern is used extremely often in the OMPL interface. It is based on the central design decision in OMPL to pass states around as pointers. Every function should cast the pointer to the proper type before using it. Because it is used so often, OMPL defines a helper method [State::<T>as()](https://ompl.kavrakilab.org/classompl_1_1base_1_1State.html#ace9774faf8793299dc105292f8557987) to make this easier.

For a constrained state space, it gets more complicated. We do not pass around `State` pointers, but a `WrapperState` (in full [ompl::base::WrapperStateSpace::StateType](https://ompl.kavrakilab.org/classompl_1_1base_1_1WrapperStateSpace_1_1StateType.html)). This type of state has a method `getState()` to access the "normal" state pointer. This "normal" state pointer is then cast again to the appropriate type. We end up with:

```C++
ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState()->as<StateType>()
```
(This has caused me a lot of headaches (I mean, interesting challenges...), and I'm still not sure I completely understand all the implications.)

## Can we make this easier?
I think OMPL and MoveIt's OMPL interface are so tightly coupled that it is difficult to avoid the `state->as<StateType>()` pattern. But there could be ways to make it easier to use. Adding the assert statement seems like a good step in that direction.

## Q&A
> First is this an external function, do you expect users to use this or is it just for some other code you wrote?

This is all internal as far as I know. The user interface (ignoring MoveGroup for a moment) looks like this:

```C++
// setup the planning context manager
ompl_interface::PlanningContextManager pcm(robot_model_, constraint_sampler_manager_);

// get a planning context for a specific planning scene and request
auto pc = pcm.getPlanningContext(planning_scene_, request, error_code, node_handle_, false);

// solve your problem
planning_interface::MotionPlanResponse res;
bool res = pc->solve(res);
```
(The tests in `test_planning_context_manager.cpp` show this in more detail.)

> Second, if ompl_state was a const& can you use the as<>() functionality?

I'm not sure, can you static cast references? The implementation goes:

```C++
template <class T>
T *as()
 {
 BOOST_CONCEPT_ASSERT((boost::Convertible<T *, State *>));
   return static_cast<T *>(this);
}
```

Also, many of the functions that take raw state pointers implement an interface imposed by OMPL. (Not the one discussed here though.)

Reply:
It looks like there is a boost assert there so we shouldn't need to add ours.  (I'm not sure what a cncept assert is).

Looking at the examples (the tests) I don't mind this interface nearly as much.  It seems like it is this way because of the interface of ompl and the intersection with the moveit stadards (the ref to the RobotState).  I think the combination of a non-const ref and a const ptr had me confused what you were trying to do here.

Reply:
I do think adding `assert(ptr != nullptr);` would be useful.  The assert I don't think we need is one that tests if the cast will work (as that is already in the as<>() function).

Reply:
The assert statements are added.